DEVSITE-47: Implementation of Algolia plugin

naydav opened the pull request:
https://jira.corp.adobe.com/browse/DEVSITE-47

icaraps commented:
@naydav the smoke tests are failing. It seems like if you don't provide ALGOLIA env parameters it wont build. Can you pls make Algolia optional in the sense that it should still build even if you don't provide it.

bdenham commented on README.md:
````diff
@@ -386,6 +386,19 @@ ADOBE_LAUNCH_SRC=https://your.adobe.launch.url.here
+ALGOLIA_INDEX_NAME=my-application-index-name
+ALGOLIA_INDEX_DRY_RUN=true # debug mode. The data will not pushed to real index.
+```
+
````
```suggestion

As previously noted, setting `ALGOLIA_INDEX_DRY_RUN` to `true` (or leaving it undefined), shows important information about the generated records in the console, without pushing data to your remote Algolia index. This is vital during search record and index development. Without the `dryRun` option, you would be wasting our purchased allotment of record changes and search operations per month. After development, when you are ready to push a site's content to Algolia, change the setting to `false` and run `yarn build`.
```

naydav replied:
Done

bdenham commented on example/.env.example:
```diff
@@ -5,3 +5,7 @@ REPO_BRANCH=main
 GATSBY_LAUNCH_SRC=
 GATSBY_LAUNCH_SRC_INCLUDE_IN_DEVELOPMENT=
+ALGOLIA_APP_ID=
+ALGOLIA_API_KEY=
```
```suggestion
ALGOLIA_SEARCH_API_KEY=
```

naydav replied:
Done

bdenham reviewed (changes requested):
## Review Summary

I made some suggestions for how to fix the smoke test failures and updated the docs to be more explicit. Also, make sure to run `yarn format`. 

### .env variables
 
On a separate PR, we should also remove all the common `.env` variables from the site `.env.example` file. I think we only need two Algolia `.env` variables within each Adobe IO site:

- ALGOLIA_INDEX_NAME=
- ALGOLIA_INDEX_DRY_RUN=

The other three Algolia variables can safely stay within the theme in a separate config file (never published to GitHub):

- ALGOLIA_APP_ID
- ALGOLIA_WRITE_API_KEY
- ALGOLIA_SEARCH_API_KEY

Doing it this way will mitigate the risk of leaking the Algolia write key and minimize the configuration that the various AdobeIO site admins have to enter. 

### Replacing ALGOLIA_INDEX_NAME with the existing REPO_NAME

We could also remove `ALGOLIA_INDEX_NAME` from the site `.env` variables by using the already existing `REPO_NAME` for the `ALGOLIA_INDEX_NAME`. My naming strategy for core product indexes is to match each product's repo name anyway. 

```
Repo name: uxp-photoshop --> Index name: uxp-photoshop
```

We'll have to figure out how repos related to the core products will specify the index they want to use, but we can figure out something else for those later. 

### Single Algolia variable for each site: DISABLE_AUTO_SEARCH_UPDATES

After doing all that, we would only have one search config variable to configure for each site:  `ALGOLIA_INDEX_DRY_RUN`, which I would rename to something friendlier and more meaningful to site admins, like `DISABLE_AUTO_SEARCH_UPDATES`.

icaraps commented:
So are these env variables: 

```
ALGOLIA_APP_ID
ALGOLIA_WRITE_API_KEY
ALGOLIA_SEARCH_API_KEY
```
common to all sites ? If yes, we would use organisation secrets to define them so they can be shared across all sites.

bdenham commented:
@icaraps, exactly. all three can be used across sites.

bdenham commented:
Just clarified some thoughts about the `.env` variables in my summary above.

icaraps commented:
> Replacing ALGOLIA_INDEX_NAME with the existing REPO_NAME

+1 if possible

> ALGOLIA_INDEX_DRY_RUN

The dry run could be by default `true` for builds via GitHub actions which are going to stage and prod and `false` for local development (eg with `npm run dev && open localhost:8000`) so site admins don't need to think about whether it should be set or not.

bdenham commented on packages/gatsby-theme-aio/algolia/query-builder.js:
```diff
@@ -0,0 +1,94 @@
+   */
+  build = (options = {}) => {
+    this.cssSelector = options.cssSelector ? options.cssSelector : 'p,li,td,code';
+    const mdxQuery = options.mdxQuery ? options.mdxQuery : `
```
```suggestion
    const mdQuery = options.mdQuery ? options.mdQuery : `
```

You can change the query name to `mdQuery` because it is not a mdxQuery. We are using the `allMarkdownRemark` query from the `gatsby-transformer-remark` plugin to get the html, so we're just pulling from plain markdown.

icaraps replied:
> So we're just pulling from plain markdown.

Is transcluding content pulled in this case ? See https://github.com/adobe/gatsby-theme-aio#embedding-markdown-documents-and-filtering-content

UXP is a strong user of transcluded content eg https://github.com/AdobeDocs/uxp-photoshop/search?q=import

naydav replied:
@icaraps 
Looks like no, transcluding content doesn't pulled to index.
I will create a separate ticket to support it.

bdenham commented on packages/gatsby-theme-aio/algolia/query-builder.js:
```diff
@@ -0,0 +1,94 @@
+
+    return [
+      {
+        query: mdxQuery,
```
```suggestion
        query: mdQuery,
```

naydav replied:
Done

naydav commented:
@bdenham 
> Replacing ALGOLIA_INDEX_NAME with the existing REPO_NAME
From one side it's good point and makes index names more clear.
From another side sometimes it's make sense to have 2 indexes for one repository, like one prod index and one for development (for example `DEVSITE-47`)
Maybe need to have someting like "index suffix"

> Single Algolia variable for each site: DISABLE_AUTO_SEARCH_UPDATES
It could be lead to question What are "search updates"?
Maybe need to rename on something as `SKIP_SEARCH_INDEXATION`  or `SKIP_SEARCH_INDEXATION_DURING_BUILD`

> When dryRun is set to true, we need to show all the indexing statistics and record objects in the console. And when dryRun is explicitly set to false, we want to publish the records to the site's Algolia index.
Unfortunately, if the option `skipIndexing` is set `false`, the indexing mechanism will not be run and `ALGOLIA_INDEX_DRY_RUN` will have no effect.
Thus, the `skipIndexing` option is responsible for starting the indexing process, and the `dryRun` option will be whether this data will be sent to Algolia

naydav commented on README.md:
```diff
@@ -386,6 +386,19 @@ ADOBE_LAUNCH_SRC=https://your.adobe.launch.url.here
 
+### .env settings for Algolia indexing
+
+- Algolia comes with a set of predefined API keys.
```
Done

naydav commented on README.md:
```diff
@@ -386,6 +386,19 @@ ADOBE_LAUNCH_SRC=https://your.adobe.launch.url.here
+### .env settings for Algolia indexing
+
+- Algolia comes with a set of predefined API keys.
+- `ALGOLIA_INDEX_DRY_RUN`. If set to `true` (or missed in .env), provides index records info in console, but doesn't push data to a real Algolia index. For pushing data should be set as `false`.
```
Done

naydav commented on README.md:
```diff
@@ -386,6 +386,19 @@ ADOBE_LAUNCH_SRC=https://your.adobe.launch.url.here
+- Algolia comes with a set of predefined API keys.
+- `ALGOLIA_INDEX_DRY_RUN`. If set to `true` (or missed in .env), provides index records info in console, but doesn't push data to a real Algolia index. For pushing data should be set as `false`.
+
+Example:
```
Done

naydav commented:
> > Replacing ALGOLIA_INDEX_NAME with the existing REPO_NAME
> 
> +1 if possible
> 
> > ALGOLIA_INDEX_DRY_RUN
> 
> The dry run could be by default `false` for builds via GitHub actions which are going to stage and prod and `true` for local development (eg with `npm run dev && open localhost:8000`) so site admins don't need to think about whether it should be set or not.

Done

bdenham commented:
> Unfortunately, if the option `skipIndexing` is set `false`, the indexing mechanism will not be run and `ALGOLIA_INDEX_DRY_RUN` will have no effect.
> Thus, the `skipIndexing` option is responsible for starting the indexing process, and the `dryRun` option will be whether this data will be sent to Algolia

No, it's the other way around. `dryRun` is ONLY executed when `skipIndexing` is false. When `skipIndexing` is true, all the indexing is SKIPPED. So the dryRun is never executed.

bdenham commented on packages/gatsby-theme-aio/gatsby-config.js:
```diff
@@ -14,6 +14,20 @@ require('dotenv').config({
+  console.error('Missed required parameter: ALGOLIA_INDEX_NAME');
+}
+const algoliaDryRun = !(process.env.ALGOLIA_INDEX_DRY_RUN && process.env.ALGOLIA_INDEX_DRY_RUN === 'false');
+
```
See comment elsewhere.

naydav commented:
@bd

> > Unfortunately, if the option `skipIndexing` is set `false`, the indexing mechanism will not be run and `ALGOLIA_INDEX_DRY_RUN` will have no effect.
> > Thus, the `skipIndexing` option is responsible for starting the indexing process, and the `dryRun` option will be whether this data will be sent to Algolia
> 
> No, it's the other way around. `dryRun` is ONLY executed when `skipIndexing` is false. When `skipIndexing` is true, all the indexing is SKIPPED. So the dryRun is never executed. In other words, the point I was trying to make is that `skipIndexing` should always be false, UNLESS the `ALGOLIA_INDEX_DRY_RUN` is `undefined`. And in such a case, the `gatsby-plugin-algolia` default settings for `skipIndexing` and `dryRun` are both false. And that means that every time one of the many adobe.io sites is built, records will be published to Algolia. And that is bad! So my suggested change is a built-in safety mechanism to prevent such scenarios when the `.env` variable is inevitably not set.

@bdenham 

Sorry, I made mistake
I meant "skipIndexing is set true"

naydav commented:
According to our conversation with Bruce:

1) Introduced `ALGOLIA_INDEXATION: skip | console | index` option
Where is:
`skip` = `skipIndexing = true, dryRun = doesn't matter`
`console` =  `skipIndexing = false, dryRun = true`
`index` =  `skipIndexing = false, dryRun = false`

2) Introduced `ALGOLIA_INDEX_SUFFIX_NAME` parameter - Optional. The name is composed of two parameters `REPO_NAME + ALGOLIA_INDEX_SUFFIX_NAME`

bdenham reviewed (approved):
Looks good! Great start to build from. Now we iterate!
